command/agent/dns.go

```diff
@@ -33,6 +33,37 @@ type DNSServer struct {
 	logger       *log.Logger
 }
 
+// Shutdown stops the DNS Servers
+func (d *DNSServer) Shutdown() error {
+	// Async stop the DNS Servers, handle a potential error
+	errCh := make(chan error, 1)
+	go func(d *DNSServer) {
+		if err := d.dnsServer.Shutdown(); err != nil {
+			d.logger.Printf("[ERR] dns: error stopping udp server: %v", err)
+			errCh <- fmt.Errorf("dns udp teardown failed: %v", err)
+		}
+	}(d)
+
+	errChTCP := make(chan error, 1)
+	go func(d *DNSServer) {
+		if err := d.dnsServerTCP.Shutdown(); err != nil {
+			d.logger.Printf("[ERR] dns: error stopping tcp server: %v", err)
+			errChTCP <- fmt.Errorf("dns tcp teardown failed: %v", err)
+		}
+	}(d)
+
+	// Wait for either the check, listen error, or timeout
```

Comment:
I see, we are not setting the `ReadTimeout` to anything in particular, so it must be defaulting to 2s. We can keep the separate routines, maybe just add that signal channel so we don't always wait 3s and timeout.